spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TemporalConvolution.scala

```diff
@@ -0,0 +1,447 @@
+ * @param wRegularizer instance of [[Regularizer]]
+ *                    (eg. L1 or L2 regularization), applied to the input weights matrices.
+ * @param bRegularizer instance of [[Regularizer]]
+ *                    applied to the bias.
+ * @param initWeight Initial weight
+ * @param initBias Initial bias
+ * @param initGradWeight Initial gradient weight
+ * @param initGradBias Initial gradient bias
+ * @tparam T The numeric type in the criterion, usually which are [[Float]] or [[Double]]
+ */
+class TemporalConvolution[T: ClassTag](
+  val inputFrameSize: Int,
+  val outputFrameSize: Int,
+  val kernelW: Int,
+  val strideW: Int = 1,
+  val propagateBack: Boolean = true,
+  var wRegularizer: Regularizer[T] = null,
+  var bRegularizer: Regularizer[T] = null,
+  val initWeight: Tensor[T] = null,
+  val initBias: Tensor[T] = null,
+  val initGradWeight: Tensor[T] = null,
+  val initGradBias: Tensor[T] = null
+)(implicit ev: TensorNumeric[T]) extends TensorModule[T]  with Initializable {
+
+  val weight: Tensor[T] = Tensor[T](outputFrameSize, inputFrameSize * kernelW)
+  val bias: Tensor[T] = Tensor[T](outputFrameSize)
+  val gradWeight: Tensor[T] = Tensor[T](outputFrameSize, inputFrameSize * kernelW)
+  val gradBias: Tensor[T] = Tensor[T](outputFrameSize)
+
+  {
+    val stdv = 1.0 / math.sqrt(kernelW * inputFrameSize)
+    val wInit: InitializationMethod = RandomUniform(-stdv, stdv)
+    val bInit: InitializationMethod = RandomUniform(-stdv, stdv)
+    setInitMethod(wInit, bInit)
+  }
+
+  @transient
+  protected var results: Array[Future[Unit]] = _
+
+  override def reset(): Unit = {
+    if (initWeight == null) {
+      weightInitMethod.init(weight, VariableFormat.GP_OUT_IN_KW_KH)
+    }
+    if (initBias == null) {
+      biasInitMethod.init(bias, VariableFormat.ONE_D)
+    }
+    zeroGradParameters()
+  }
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    // Require input of 2 dimensions or 3 dimensions
+    // 2d input format: time x feature
+    // 3d input format: batch x time x feature
+    require(input.dim() == 2 || input.dim() == 3,
+      "TemporalConvolution: 2D or 3D(batch mode) tensor expected for input, " +
+        s"but got ${input.dim()}")
+    // Require input to be contiguous
+    require(input.isContiguous())
+
+    var dimSeq = 1
+    var dimFeat = 2
+
+    if (input.dim() == 3) {
+      dimSeq = 2
+      dimFeat = 3
+    }
+
+    val nInputFrame = input.size(dimSeq)
+    var nOutputFrame = (nInputFrame - kernelW) / strideW + 1
+    val inputWindow = Tensor[T]()
+    var outputWindow = Tensor[T]()
+
+    // Shape check on input with inputFrameSize and kernelW
+    require(input.size(dimFeat) == inputFrameSize, "Invalid input frame size. Got: " +
+      s"${input.size(dimFeat)}, Expected: $inputFrameSize")
+    require(nOutputFrame >= 1, "Input sequence smaller than kernel size. Got: " +
+      s"$nInputFrame, Expected: $kernelW")
+
+    val weightT = weight.transpose(1, 2)
```

Comment:
Transpose would create a new tensor, weight will update each iteration, so we need to get a new transpose each iteration, is it right?